src/main/java/org/araymond/joal/core/ttorrent/client/ConnectionHandler.java

```diff
@@ -12,11 +11,22 @@
 import java.io.InputStreamReader;
 import java.net.*;
 import java.nio.channels.ServerSocketChannel;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 
+import static java.util.concurrent.TimeUnit.MINUTES;
+
 /**
+ * This class has 2 main functions:
+ * <ul>
+ *     <li>establishes a socket at a port that accepts connections</li>
+ *     <li>periodically resolves our external IP address</li>
+ * </ul>
+ * Note this port & IP will be reported back to trackers via announcements, if your client
+ * {@code query} contains relevant placeholder(s).
+ * <p/>
```

Comment:
Is `ConnectionHandler` really needed? As in do we _really_ need to create a socket and listen to incoming connectinos?

Reply:
Yes we definitely need it.
But it has been a work in progress since the begening... i've never finished this part.

Reply:
It's not a "ban reason" for sure, but i may be suspicous if you generate upload while not being accessibile from the outside world.

To be able to seed you need to have an open port on your router, there is no way around. Unless there is some part of the BitTorrent protocol that allow another way (maybe the client can send random parts to randiom peers on his behalf), i don't know the protocol very much on this part, i've never took time to fully RTFM ^^.

To have this port open, most BitTorrent client use UPnP to automatically create a port forwarding rule on your routers. This allow incomming traffic from the internet to reach your client.

So this is an important part of the code. It's not fully implemented (ideally you would listen for request and answer with "choke"), but it will do the trick if you manually create a port forwarding rule on your router.